### PR TITLE
Update guidance on security headers for REST APIs

### PR DESCRIPTION
**Pull request by rbsec**
As per the discussion in #445.

### CODE REVIEW

File: cheatsheets/REST_Security_Cheat_Sheet.md [hunk 1]
```diff
@@ -131,11 +131,27 @@ Services including script code (e.g. JavaScript) in their responses must be espe
...
-## Security headers
+## Security Headers
 
-To make sure the content of a given resources is interpreted correctly by the browser, the server should always send the `Content-Type` header with the correct content type, and preferably the `Content-Type` header should include a charset. The server should also send the `X-Content-Type-Options: nosniff` [security header](https://owasp.org/www-project-secure-headers/) to make sure the browser does not try to detect a different `Content-Type` than what is actually sent (can lead to XSS).
+There are a number of [security related headers](https://owasp.org/www-project-secure-headers/) that can be returned in the HTTP responses to instruct browsers to act in specific ways. However, some of these headers are intended to be used with HTML responses, and as such may provide little or no security benefits on an API that does not return HTML.
 
-Additionally the server should send the `X-Frame-Options: deny` [security header](https://owasp.org/www-project-secure-headers/) to protect against drag'n drop clickjacking attacks in older browsers.
+The following headers should be included in all API responses:
+
+| Header | Rationale |
```
Comment (ThunderSon) [inline cheatsheets/REST_Security_Cheat_Sheet.md:140]: Part of these should be the `Cache-Control`, as it's important to be private without caching for user data.

File: cheatsheets/REST_Security_Cheat_Sheet.md [hunk 2]
```diff
@@ -131,11 +131,27 @@ Services including script code (e.g. JavaScript) in their responses must be espe
...
+|--------|-----------|
+| `Content-Security-Policy: frame-ancestors 'none'` | To protect against drag-and-drop style clickjacking attacks. |
```
Comment (ThunderSon) [inline cheatsheets/REST_Security_Cheat_Sheet.md:142]: This is not required if HTML is not returned, or any "functionality" that could be triggered, so framing protection should be part of the second table, no?

Comment (rbsec) [reply cheatsheets/REST_Security_Cheat_Sheet.md:142]: There's some interesting variants on clickjacking where you get to the user to select text and then drag it into a text box (that you control) so you can read the contents. That could still potentially be exploitable against an endpoint that returns JSON/text:

https://www.w3.org/Security/wiki/Clickjacking_Threats#Drag_and_drop_attacks

File: cheatsheets/REST_Security_Cheat_Sheet.md (hunk 1 above)
Comment (rbsec) [reply cheatsheets/REST_Security_Cheat_Sheet.md:140]: Good call, added in 84acb9f4.

**Comment by ThunderSon** [comment]
@dotnetnate does this answer your issue? :)

**Comment by mackowski** [comment]
@rbsec please do not merge it yet. I would like to review it to check if we are not missing any case.

File: cheatsheets/REST_Security_Cheat_Sheet.md [hunk 3]
```diff
@@ -131,11 +131,28 @@ Services including script code (e.g. JavaScript) in their responses must be espe
...
+| `X-Content-Type-Options: nosniff` | To prevent browsers from performing MIME sniffing, and inappropriately interpreting responses as HTML. |
+| `X-Frame-Options: DENY` | To protect against drag-and-drop style clickjacking attacks. |
+
+The headers below are only intended to provide additional security when responses are rendered as HTML. As such, if the API will __never__ return HTML in responses, then these headers may not be necessary. However, if there is any uncertainty about the function of the headers, or the types of information that the API returns (or may return in future), then it is recommended to include them as part of a defence-in-depth approach.
+
+| Header | Rationale |
+|--------|-----------|
+| `Content-Security-Policy: default-src 'none'` | The majority of CSP functionality only affects pages rendered as HTML. |
+| `Feature-Policy: 'none'` | Feature policies only affect pages rendered as HTML. |
+| `Referred-Policy: no-referrer` | Non-HTML responses should not trigger additional requests. |
```
Comment (koenrh) [inline cheatsheets/REST_Security_Cheat_Sheet.md:155]:
```suggestion
| `Referrer-Policy: no-referrer` | Non-HTML responses should not trigger additional requests. |
```

**Pull request closed without merging**